File: suites/08_media_encoding_kernels/comparisons/webp_encode.cpp
#include <iostream>
#include <vector>
#include <cstdlib>

inline int64_t paeth_predictor(int64_t left, int64_t top, int64_t top_left) {
    int64_t p = left + top - top_left;
    int64_t pa = std::abs(p - left);
    int64_t pb = std::abs(p - top);
    int64_t pc = std::abs(p - top_left);

    if (pa <= pb && pa <= pc) return left;
    if (pb <= pc) return top;
    return top_left;
}

int64_t webp_encode(int64_t width, int64_t height) {
    const int64_t mod_byte = 256;
    const int64_t mod_prime = 1000000007;

    int64_t residual_sum = 0;
    for (int64_t y = 1; y < height; ++y) {
        for (int64_t x = 1; x < width; ++x) {
            int64_t left = (((x - 1) * 17) + (y * 23)) % mod_byte;
            int64_t top = ((x * 17) + ((y - 1) * 23)) % mod_byte;
            int64_t top_left = (((x - 1) * 17) + ((y - 1) * 23)) % mod_byte;
            int64_t current = ((x * 17) + (y * 23)) % mod_byte;

            int64_t predicted = paeth_predictor(left, top, top_left);
            int64_t diff = std::abs(current - predicted);
            residual_sum = (residual_sum + diff) % mod_prime;
        }
    }
    return residual_sum;
}

int main() {
    int64_t res = webp_encode(512, 512);
    std::cout << res << std::endl;
    return 0;
}

File: suites/08_media_encoding_kernels/comparisons/webp_encode.rs
fn paeth_predictor(left: i64, top: i64, top_left: i64) -> i64 {
    let p = left + top - top_left;
    let pa = (p - left).abs();
    let pb = (p - top).abs();
    let pc = (p - top_left).abs();

    if pa <= pb && pa <= pc {
        left
    } else if pb <= pc {
        top
    } else {
        top_left
    }
}

fn webp_encode(width: i64, height: i64) -> i64 {
    let mod_byte: i64 = 256;
    let mod_prime: i64 = 1000000007;

    let mut residual_sum: i64 = 0;
    for y in 1..height {
        for x in 1..width {
            let left = (((x - 1) * 17) + (y * 23)) % mod_byte;
            let top = ((x * 17) + ((y - 1) * 23)) % mod_byte;
            let top_left = (((x - 1) * 17) + ((y - 1) * 23)) % mod_byte;
            let current = ((x * 17) + (y * 23)) % mod_byte;

            let predicted = paeth_predictor(left, top, top_left);
            let diff = (current - predicted).abs();
            residual_sum = (residual_sum + diff) % mod_prime;
        }
    }
    residual_sum
}

fn main() {
    let res = webp_encode(512, 512);
    println!("{}", res);
}
